keps/2941-DRA-Structured-Parameters/README.md

```diff
@@ -0,0 +1,394 @@
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Resource Quota API
+
+```golang
+type ResourceQuota struct {
+  // ...
+	// Kind is the type of resource that this resource is
+	// +kubebuilder:validation:Enum={Core,DeviceClass}
+	// +kubebuilder:default=Core
+	Kind ResourceKind `json:"kind"`
+}
+```
+
+Kind allows one to distinguish between a Core resource and a Device class.
+
+With this, a cluster queue could be defined as follows:
+
+```yaml
+apiVersion: kueue.x-k8s.io/v1beta1
+kind: ClusterQueue
+metadata:
+  name: "cluster-queue"
+spec:
+  namespaceSelector: {} # match all.
+  resourceGroups:
+  - coveredResources: ["cpu", "memory", "gpu.example.com"]
+    flavors:
+    - name: "default-flavor"
+      resources:
+      - name: "cpu"
+        nominalQuota: 9
+      - name: "memory"
+        nominalQuota: "200Mi"
+      - name: "gpu.example.com"
+        nominalQuota: 2
+        kind: "DeviceClass"
+```
+
+### Workloads
+
+When a user submits a workload and KueueDynamicResourceAllocation feature gate is on, Kueue will do the following:
+
+a. Claims will be read from resources.claims in the PodTemplateSpec.
+b. The name of the claim will be used to look up the corresponding `ResourceClaimTemplateName` in the PodTemplateSpec.
+c. The ResourceClaim will be read given the name in b and using the same namespace as the workload.
+d. From the ResourceClaimTemplate, the deviceClassName will be read.
+e. Every claim that requests the same deviceClassName will be tallied and reported in the ResourceUsage.
+
+```yaml
+---
+
+apiVersion: batch/v1
+kind: Job
+metadata:
+  namespace: gpu-test1
+  name: job0
+  labels:
+    app: job
+    kueue.x-k8s.io/queue-name: user-queue
+spec:
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: ctr0
+        image: ubuntu:22.04
+        command: ["bash", "-c"]
+        args: ["export; sleep 9999"]
+        resources:
+          claims:
+          - name: gpu. #a) read the claim from resources.claims
+          requests:
+            cpu: 1
+            memory: "200Mi"
+      resourceClaims:
+      - name: gpu # b) use the name in resources.claim
+        resourceClaimTemplateName: single-gpu # c) the name for resource claim templates 
+---
+apiVersion: resource.k8s.io/v1alpha3
+kind: ResourceClaimTemplate
+metadata:
+  namespace: gpu-test1
+  name: single-gpu
+spec:
+  spec:
+    devices:
+      requests:
+      - name: gpu
+        deviceClassName: gpu.example.com # d) the name of the device class
```

Comment:
Would this still count against the quota if there was also some additional filtering criteria in the request's `selectors` that prevent kube-scheduler from fulfilling the request?

Reply:
> Would this still count against the quota if there was also some additional filtering criteria in the request's selectors that prevent kube-scheduler from fulfilling the request?

Yes, IMHO, this is no different than device plugins based device being counted against quota but scheduler not able to land the pod on a node due to missing out of other scheduling constraints.